content/blog/2022-05-05-e2e-cv-pt3.md

```diff
@@ -0,0 +1,292 @@
+              heroku_email: 'alexkim@iterative.ai'
+              team: 'iterative-sandbox'
+              usedocker: true
+      ```
+
+      Now we can test our API and assess the end-to-end performance of the
+      overall solution. If we pick a random test image `exp3_num_258558.jpg`,
+      ![test image `exp3_num_258558.jpg`](/uploads/images/2022-05-05/exp3_num_258558.jpg '=300')
+
+      we can send it to the endpoint using the `curl` command like this:
+
+      ```bash
+      $ curl -F 'image=@data/MAGNETIC_TILE_SURFACE_DEFECTS/test_images/exp3_num_258558.jpg' \
+      -v https://demo-api-mag-tiles-dev.herokuapp.com/analyze
+      ```
+
+      This will return some http-header info and the body of the response
+      containing the defect segmentation mask (`0` for pixel locations without
+      defects and `1` otherwise):
+
+      ```json
+      *   Trying 18.208.60.216:443...
+      * Connected to demo-api-mag-tiles-dev.herokuapp.com (18.208.60.216) port 443 (#0)
+      ...
+      {"pred":[[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
+      0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
+      0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
+      1,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
+      ...
+      ```
+
+      Alternatively, we can do a similar thing with a Python script that also
+      saves the output mask into a `exp3_num_258558_mask.png` image:
+
+      ```python
+      import json
+      from pathlib import Path
+
+      import matplotlib.cm as cm
+      import matplotlib.pyplot as plt
+      import numpy as np
+      import requests
+
+      url = 'https://demo-api-mag-tiles-dev.herokuapp.com/analyze'
+      file_path = Path(
+          'data/MAGNETIC_TILE_SURFACE_DEFECTS/test_images/exp3_num_258558.jpg')
+      files = {'image': (str(file_path), open(file_path, 'rb'), "image/jpeg")}
+      response = requests.post(url, files=files)
+      data = json.loads(response.content)
+      pred = np.array(data['pred'])
+      plt.imsave(f'{file_path.stem}_mask.png', pred, cmap=cm.gray)
+      ```
+
+      Below you can see what this mask looks like.
+      ![Output mask `exp3_num_258558_mask.png`](/uploads/images/2022-05-05/exp3_num_258558_mask.png '=300')
+
+3. [Workflow for deploying to the production environment](https://github.com/iterative/magnetic-tiles-defect/blob/main/.github/workflows/3-deploy.yaml)
+
+   ![Workflow for deploying to the production environment](/uploads/images/2022-05-05/workflow_prod.png '=800')
+
+   Before we merge the dev branch into the main branch, we would need to
+   thoroughly test and monitor the application in the development environment. A
+   good test could be duplicating real image requests to the dev endpoint (here
+   it's
+   [https://demo-api-mag-tiles-**dev**.herokuapp.com/analyze](https://demo-api-mag-tiles-dev.herokuapp.com/analyze))
```

Comment:
not sure I like this link showing, but also not super opposed in this circumstance.  Maybe the link would be in "dev endpoint?"  Also why is 'dev' bolded in the link?